dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -465,131 +523,122 @@ impl<'a, 'tcx: 'a> Visitor<'tcx> for CollectFunctionInstrumentationPoints<'a, 't
-                                source: p.place().as_ref().map(to_mir_place),
-                                destination: Some(to_mir_place(&dest)),
-                                transfer_kind: TransferKind::None,
-                            },
-                        );
-                    } else {
-                        // Casting between pointer types; trace the destination
+                    Rvalue::Cast(_, p, _) => {
+                        location.statement_index += 1;
+                        if p.ty(&self.body.local_decls, self.tcx).is_integral() {
+                            // Casting integer to a pointer; trace the destination
+                            self.add_instrumentation_point(
+                                location,
+                                ptr_contrive_fn,
+                                vec![InstrumentationOperand::from_type(
+                                    Operand::Copy(dest),
+                                    &dest_ty,
+                                )],
+                                false,
+                                false,
+                                EventMetadata {
+                                    source: p.place().as_ref().map(to_mir_place),
+                                    destination: Some(to_mir_place(&dest)),
+                                    transfer_kind: TransferKind::None,
+                                },
+                            );
+                        } else {
+                            // Casting between pointer types; trace the destination
+                            self.add_instrumentation_point(
+                                location,
+                                copy_fn,
+                                vec![InstrumentationOperand::from_type(
+                                    Operand::Copy(dest),
+                                    &dest_ty,
+                                )],
+                                false,
+                                false,
+                                EventMetadata {
+                                    source: p.place().as_ref().map(to_mir_place),
+                                    destination: Some(to_mir_place(&dest)),
+                                    transfer_kind: TransferKind::None,
+                                },
+                            );
+                        }
+                    }
+                    Rvalue::AddressOf(_, p) => {
+                        // Instrument which local's address is taken
+                        location.statement_index += 1;
                         self.add_instrumentation_point(
                             location,
-                            copy_fn,
-                            vec![InstrumentationOperand::RawPtr(Operand::Copy(dest))],
+                            addr_local_fn,
+                            vec![
+                                InstrumentationOperand::from_type(Operand::Copy(dest), &dest_ty),
```

Comment:
The type of this destination is always a raw ptr because `AddressOf` signifies "`&v as *const _` or raw address of expressions like `&raw v` or `addr_of!(v)`", so don't use `from_type` here to dynamically dispatch on the statically-known type.